File: leetcode_ly/c++/Tree/FlattenBinaryTreeToLinkedList.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if(!root) return;
        while(root){
            if(root->left){
                TreeNode* node = root->left;
                while(node->right){
                    node = node->right;
                }
                node->right = root->right;
                root->right = root->left;
                root->left = NULL;
            }
            root = root->right;
        }
    }
};

File: leetcode_ly/java/Tree/FlattenBinaryTreeToLinkedList.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        if(root == null) return;
        while(root.right != null && root.left == null) root = root.right;
        while(root.left != null){
            TreeNode node = root.left;
            TreeNode ptr = root.right;
            root.right = node;
            root.left = null;
            while(node.right != null){
                node = node.right;
            }
            node.right = ptr;
            while(root.right != null && root.left == null) root = root.right;
        }
        return;
    }
}